nomad/drainer/drain_testing.go

```diff
@@ -1,52 +1,142 @@
-	Nodes  map[string]*structs.Node
-	Events []*MockNodeTrackerEvent
+type MockJobWatcher struct {
+	drainCh    chan *DrainRequest
+	migratedCh chan []*structs.Allocation
+	jobs       map[structs.NamespacedID]struct{}
 	sync.Mutex
 }
 
-func NewMockNodeTracker() *MockNodeTracker {
-	return &MockNodeTracker{
-		Nodes:  make(map[string]*structs.Node),
-		Events: make([]*MockNodeTrackerEvent, 0, 16),
+// RegisterJobs marks the job as being watched
+func (m *MockJobWatcher) RegisterJobs(jobs []structs.NamespacedID) {
+	m.Lock()
+	defer m.Unlock()
+	for _, job := range jobs {
+		m.jobs[job] = struct{}{}
 	}
 }
 
-func (m *MockNodeTracker) TrackedNodes() map[string]*structs.Node {
-	m.Lock()
-	defer m.Unlock()
-	return m.Nodes
+// Drain returns the DrainRequest channel. Tests can send on this channel to
+// simulate steps through the NodeDrainer watch loop. (Sending on this channel
+// will block anywhere else.)
+func (m *MockJobWatcher) Drain() <-chan *DrainRequest {
+	return m.drainCh
 }
 
-func (m *MockNodeTracker) Remove(nodeID string) {
-	m.Lock()
-	defer m.Unlock()
-	delete(m.Nodes, nodeID)
-	m.Events = append(m.Events, &MockNodeTrackerEvent{NodeRemove: nodeID})
+// Migrated returns the channel of migrated allocations. Tests can send on this
+// channel to simulate steps through the NodeDrainer watch loop. (Sending on
+// this channel will block anywhere else.)
+func (m *MockJobWatcher) Migrated() <-chan []*structs.Allocation {
+	return m.migratedCh
+}
+
+type MockDeadlineNotifier struct {
+	expiredCh <-chan []string
+	nodes     map[string]struct{}
+	sync.Mutex
 }
 
-func (m *MockNodeTracker) Update(node *structs.Node) {
+// NextBatch returns the channel of expired nodes. Tests can send on this
+// channel to simulate timer events in the NodeDrainer watch loop. (Sending on
+// this channel will block anywhere else.)
+func (m *MockDeadlineNotifier) NextBatch() <-chan []string {
+	return m.expiredCh
+}
+
+// Remove removes the given node from being tracked for a deadline.
+func (m *MockDeadlineNotifier) Remove(nodeID string) {
 	m.Lock()
 	defer m.Unlock()
-	m.Nodes[node.ID] = node
-	m.Events = append(m.Events, &MockNodeTrackerEvent{NodeUpdate: node})
+	delete(m.nodes, nodeID)
 }
 
-func (m *MockNodeTracker) events() []*MockNodeTrackerEvent {
+// Watch marks the node as being watched; this mock throws out the timer in lieu
+// of manully sending on the channel to avoid racy tests.
+func (m *MockDeadlineNotifier) Watch(nodeID string, _ time.Time) {
 	m.Lock()
 	defer m.Unlock()
+	m.nodes[nodeID] = struct{}{}
+}
+
+type MockRaftApplierShim struct {
+	state *state.StateStore
+}
+
+// AllocUpdateDesiredTransition mocks a write to raft as a state store update
+func (m *MockRaftApplierShim) AllocUpdateDesiredTransition(
+	allocs map[string]*structs.DesiredTransition, evals []*structs.Evaluation) (uint64, error) {
+	index, _ := m.state.LatestIndex()
+	index++
+	err := m.state.UpdateAllocsDesiredTransitions(structs.MsgTypeTestSetup, index, allocs, evals)
```

Comment:
I think this needs locking to ensure data races can't cause bad interleaving with concurrent index reads and increments.

Reply:
Good catch, will fix.